bochs/iodev/usb/scsi_device.cc

```diff
@@ -1181,6 +1384,343 @@ Bit32s scsi_device_t::scsi_send_command(Bit32u tag, Bit8u *buf, Bit8u cmd_len, i
+        p[17] = (0x00 >>  0) & 0xFF;   // IDLE_C condition timer (LSB)
+        p[18] = (0x00 >> 24) & 0xFF;   // STANDBY_Y condition timer (MSB) (valid if STANDBY_Y = 1)
+        p[19] = (0x00 >> 16) & 0xFF;   // STANDBY_Y condition timer    (100 mS increments)
+        p[20] = (0x00 >>  8) & 0xFF;   // STANDBY_Y condition timer
+        p[21] = (0x00 >>  0) & 0xFF;   // STANDBY_Y condition timer (LSB)
+        memset(&p[22], 0, 16);         // reserved
+      }
+      size += 38;
+      break;
+      
+    case PAGE_EXCEPTION_CONTROL: // Information Exceptions Control page
+      size = scsi_do_modepage_hdr(p, sub_page, page_code, 10);
+      p += size;
+      if (pc == PAGE_CONTROL_CHANGEABLE) {
+        memset(p, 0, 10);
+      } else {
+        p[ 0] = 0;                    // PERF  EBF  EWASC  DEXCPT  TEST  EBACKERR  LOGERR
+        p[ 1] = 0;                    // MRIE
+        p[ 2] = (0  >> 24) & 0xFF;    // interval timer (MSB)
+        p[ 3] = (0  >> 16) & 0xFF;    // interval timer   (0x00000000 = vendor specific)
+        p[ 4] = (0  >>  8) & 0xFF;    // interval timer
+        p[ 5] = (0  >>  0) & 0xFF;    // interval timer (LSB)
+        p[ 6] = (0  >> 24) & 0xFF;    // report count (MSB)
+        p[ 7] = (0  >> 16) & 0xFF;    // report count   (0x00000000 = no limit)
+        p[ 8] = (0  >>  8) & 0xFF;    // report count
+        p[ 9] = (0  >>  0) & 0xFF;    // report count (LSB)
+      }
+      size += 10;
+      break;
+      
+    case PAGE_CDVD_INACTIVITY: // CD-ROM inactivity page
+      if (type == SCSIDEV_TYPE_CDROM) {
+        size = scsi_do_modepage_hdr(p, sub_page, page_code, 10);
+        p += size;
+        if (pc == PAGE_CONTROL_CHANGEABLE) {
+          memset(p, 0, 10);
+        } else {
+          p[ 0] = 0;                    // reserved
+          p[ 1] = 0;                    // reserved
+          p[ 2] = 0;                    // G3Enable  TMOE  DISP  SWPP
+          p[ 3] = 0;                    // reserved
+          p[ 4] = (0  >>  8) & 0xFF;    // Group 1 Min Timeout (seconds) (MSB) (valid only if TMOE = 1)
+          p[ 5] = (0  >>  0) & 0xFF;    // Group 1 Min Timeout (seconds) (LSB)
+          p[ 6] = (0  >>  8) & 0xFF;    // Group 2 Min Timeout (seconds) (MSB)
+          p[ 7] = (0  >>  0) & 0xFF;    // Group 2 Min Timeout (seconds) (LSB)
+          p[ 8] = (0  >>  8) & 0xFF;    // Group 3 Timeout (milliseconds) (MSB) (valid only if G3Enable = 1)
+          p[ 9] = (0  >>  0) & 0xFF;    // Group 3 Timeout (milliseconds) (LSB)
+        }
+        size += 10;
+      }
+      break;
+      
+    // MMC3r10g, section 6.3.11, page 311(347)
+    case PAGE_CAPABILITIES:  // capabilities page
+      if (type == SCSIDEV_TYPE_CDROM) {
+        size = scsi_do_modepage_hdr(p, sub_page, page_code, 24);
+        p += size;
+        if (pc == PAGE_CONTROL_CHANGEABLE) {
+          memset(p, 0, 24);
+        } else {
+          p[ 0] = 0x03;                 // CD-R or CD-RW read/only
```

Comment:
Ben, for future - I think it could be good idea to have defined struct (properly packed and etc) with all relevant fields, fill the struct on the side filed-by-field and then use memcpy to put into the buffer p

Reply:
I agree, though the reason I did it this way, for now, is that some of the fields may change depending on the device it is associated with. If, later, I find that the fields will remain static, I can do as you suggest.  Even if only one of the fields change, as with this example, it will still be beneficial. Thank you.